panic: Failed to create memberlist

<!-- Issue template; please answer the questions. -->

**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind feature

**What happened**:

```
log file: /tmp/memberlist.log
panic: Failed to create memberlist: Failed to get final advertise address: No private IP address found, and explicit IP not provided

goroutine 17 [running, locked to thread]:
main.init_memberlist(0x7fe8cbf90600, 0x0)
	/root/ftlib/ftlib/consensus/gossip/memberlist/main.go:41 +0x245
main._cgoexpwrap_fc57e92fce6d_init_memberlist(0x7fe8cbf90600, 0x11090e0)
	_cgo_gotypes.go:78 +0x2d
```
**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:
* Clone https://github.com/sql-machine-learning/elasticdl/tree/allreduce-test
* Run the following command
```
docker build --target dev -t elasticdl:dev -f elasticdl/docker/Dockerfile .
docker build --target ci -t elasticdl:ci -f elasticdl/docker/Dockerfile .
docker build -t elasticdl:dev_allreduce -f elasticdl/docker/Dockerfile.dev_allreduce .

python3 -m elasticdl.python.elasticdl.client train \
  --image_base=elasticdl:dev_allreduce \
  --docker_image_repository=reg.docker.replace-with-your-registry-here.com/yuantang/ \
  --model_zoo=model_zoo \
  --model_def=deepfm_functional_api.deepfm_functional_api.custom_model \
  --training_data=/data/frappe/train \
  --validation_data=/data/frappe/test \
  --num_epochs=1 \
  --master_resource_request="cpu=1,memory=4096Mi,ephemeral-storage=10240Mi" \
  --master_resource_limit="cpu=1,memory=4096Mi,ephemeral-storage=10240Mi" \
  --worker_resource_request="cpu=1,memory=4096Mi,ephemeral-storage=10240Mi" \
  --worker_resource_limit="cpu=1,memory=4096Mi,ephemeral-storage=10240Mi" \
  --minibatch_size=64 \
  --num_minibatches_per_task=2 \
  --num_workers=2 \
  --checkpoint_steps=500 \
  --evaluation_steps=500 \
  --grads_to_wait=1 \
  --job_name=test-train \
  --log_level=INFO \
  --image_pull_policy=Always \
  --output=/saved_model/model_output \
  --volume="host_path=${PWD},mount_path=/saved_model" \
   --namespace=<replace-by-your-namespace-here> \
  --distribution_strategy=AllreduceStrategy

kub logs elasticdl-test-train-worker-1
```
**Anything else we need to know?**:
Relevant code:
* This is how we wrap FTLib for ElasticDL to use: https://github.com/sql-machine-learning/elasticdl/blob/develop/elasticdl/python/collective_ops/communicator.py
* This is where the wrapped FTLib ops are being used: https://github.com/sql-machine-learning/elasticdl/blob/develop/elasticdl/python/worker/worker.py#L793

I am seeing a different error when testing on Minikube on Travis: https://travis-ci.org/github/sql-machine-learning/elasticdl/builds/666366309
```
log file: /tmp/memberlist.log
joining with ips:
10.111.2.235
panic: Failed to join cluster: 2 errors occurred:
	* Failed to resolve : lookup : no such host
	* Failed to join 10.111.2.235: dial tcp 10.111.2.235:7946: connect: connection refused
goroutine 17 [running, locked to thread]:
main.join(0x7fc730032450, 0x1, 0x1, 0x7fc6eb4a6c98)
	/root/ftlib/ftlib/consensus/gossip/memberlist/main.go:57 +0x335
main._cgoexpwrap_fc57e92fce6d_join(0x7fc730032450, 0x1, 0x1, 0xc)
	_cgo_gotypes.go:91 +0x41
```

--- a/ftlib/consensus/gossip/impl.py
+++ b/ftlib/consensus/gossip/impl.py
@@ -14,7 +14,12 @@ from ftlib.consensus.consensus_status import ConsensusMode, ConsensusStatus
 #####################################################################
 class Gossip(BasicConsensus):
     def __init__(
-        self, ftlib, known_addr_list, log_file="/tmp/memberlist.log",
+        self,
+        ftlib,
+        known_addr_list,
+        log_file="/tmp/memberlist.log",
+        custom_bind_addr="",
+        custom_advertise_addr="",
     ):
         super(Gossip, self).__init__()
 
@@ -30,7 +35,11 @@ class Gossip(BasicConsensus):
         self._lib.join.argtypes = [self._goslice_type]
         self._lib.get_memberlist.restype = self.member_list
 
-        res = self._lib.init_memberlist(log_file.encode("utf-8"))
+        res = self._lib.init_memberlist(
+            log_file.encode("utf-8"),
+            custom_bind_addr.encode("utf-8"),
+            custom_advertise_addr.encode("utf-8"),
+        )
         if res != 0:
             raise RuntimeError("failed to initialize memberlist")
 

--- a/ftlib/consensus/gossip/memberlist/main.go
+++ b/ftlib/consensus/gossip/memberlist/main.go
@@ -21,11 +21,25 @@ func main() {
 }
 
 //export init_memberlist
-func init_memberlist(cLogFileName *C.char) C.int {
+func init_memberlist(
+    cLogFileName *C.char,
+    cBindAddr *C.char,
+    cAdvertiseAddr *C.char,
+    ) C.int {
     var err error
     config := memberlist.DefaultLocalConfig()
 
     logFileName := C.GoString(cLogFileName)
+    customBindAddr := C.GoString(cBindAddr)
+    customAdvertiseAddr := C.GoString(cAdvertiseAddr)
+
+    if customBindAddr != "" {
+        config.BindAddr = customBindAddr
+    }
+
+    if customAdvertiseAddr != "" {
+        config.AdvertiseAddr = customAdvertiseAddr
+    }
 
     if logFileName != "" {
         fmt.Printf("log file: %s\n", logFileName)

